crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -185,293 +519,303 @@ where
-    use assert_matches::assert_matches;
-    use futures_util::stream::{StreamExt, TryStreamExt};
-    use reth_eth_wire::BlockBody;
-    use reth_interfaces::{
-        p2p::{bodies::downloader::BodyDownloader, error::RequestError},
-        test_utils::TestConsensus,
-    };
-    use reth_primitives::{Header, PeerId, H256};
-    use std::{
-        sync::{
-            atomic::{AtomicUsize, Ordering},
-            Arc,
-        },
-        time::Duration,
-    };
-
-    // Check that the blocks are emitted in order of block number, not in order of
-    // first-downloaded
-    #[tokio::test]
-    async fn emits_bodies_in_order() {
-        // Generate some random blocks
-        let (headers, mut bodies) = generate_bodies(0..20);
-
-        let downloader = ConcurrentDownloader::new(
-            Arc::new(TestBodiesClient::new(|hashes: Vec<H256>| {
-                let mut bodies = bodies.clone();
-                async move {
-                    // Simulate that the request for this (random) block takes 0-100ms
-                    tokio::time::sleep(Duration::from_millis(hashes[0].to_low_u64_be() % 100))
-                        .await;
-
-                    Ok((
-                        PeerId::default(),
-                        hashes
-                            .into_iter()
-                            .map(|hash| {
-                                bodies
-                                    .remove(&hash)
-                                    .expect("Downloader asked for a block it should not ask for")
-                            })
-                            .collect(),
-                    )
-                        .into())
+impl<B, C> Future for BodiesRequestFuture<B, C>
+where
+    B: BodiesClient + 'static,
+    C: ConsensusTrait + 'static,
+{
+    type Output = Vec<BlockResponse>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        loop {
+            if let Some(fut) = this.fut.as_mut() {
+                match ready!(fut.poll_unpin(cx)) {
+                    Ok(response) => {
+                        let (peer_id, bodies) = response.split();
+                        if bodies.is_empty() {
+                            this.on_error(BodyRequestError::EmptyResponse, Some(peer_id));
+                            continue
+                        }
+
+                        let request_len = this.hashes_to_download.len();
+                        let response_len = bodies.len();
+                        if response_len > request_len {
+                            this.on_error(
+                                BodyRequestError::TooManyBodies {
+                                    expected: request_len,
+                                    received: response_len,
+                                },
+                                Some(peer_id),
+                            );
+                            continue
+                        }
```

Comment:
There are malicious peers which respond with a single block. Currently, it's acceptable for the body DL to receive less blocks than expected. This means that a single body response will be accepted and the malicious peer will not be penalized. 
Possible solutions: 
1. Expect at least half of requested bodies - this is not ideal and might become a footgun if the user changes the default request batch size.
2. Calculate the response size and discard if it is below a certain threshold (1mb)?

Reply:
As long as the body validates correctly, I think it is ok that they respond with fewer bodies than we requested, but I agree we should set some minimum response size in terms of bodies responded with. I'm not sure where the footgun would be if the batch size is changed? I guess if you make the batch size too big, but that is a separate issue (that we should track) around validating user-specified parameters being in acceptable ranges

Reply:
@onbjerg your statement is correct from a data validity perspective, however, there is an issue download/sync speed. Hypothetically, a malicious peer can send back bodies one by one just before the request timeout. Then it would take `request_limit * ...` let's say `(timeout - 1)` seconds to resolve a single request, which is far from ideal.

Reply:
We should probably just penalize the peer for sending too few (e.g. less than half of requested)? I don't think we should error per se, at some point if the peer doesn't behave we will disconnect.